### DISCUSSION TITLE
Q&A: Problem about handling errors raised by async selector

### DISCUSSION THREAD

**Discussion by babyfish-ct** [facebookexperimental/Recoil#998] [Q&A]
Hi
I met a problem when I tried to handle the errors raised by async selector, I asked this problem on stack overflow, the link is [https://stackoverflow.com/questions/67251358/recoil-cannot-handle-the-error-of-async-selector](https://stackoverflow.com/questions/67251358/recoil-cannot-handle-the-error-of-async-selector)

No reply until now, can you help me? Sorry for interrupting :)

**Comment by drarmstr** [comment]
Please inline the question here.  Yes, async selectors can throw errors and if nothing catches them it is expected to the application to error.  You can use `<ErrorBoundary/>` to catch the errors, but the error boundary must be wrapping the component which uses the hook which throws the error.

**Comment by babyfish-ct** [comment]
**My Code**(by using useRecoilValueLoadable)
```
import {RecoilRoot, selector, useRecoilValueLoadable} from 'recoil';

const asyncError = selector({
  key: 'asyncError',
  get: async() => { throw new Error("Test Error"); }
});

function Test() {
  const loadable = useRecoilValueLoadable(asyncError);
  return (
    <h1>
      { loadable?.state === 'hasError' ? "Has error" : "No error" }
    </h1>
  );
}

function App() {
  return (
    <RecoilRoot>
      <Test/>
    </RecoilRoot>
  );
}

export default App;
```

**Result:**
Unhandled rejection error

I tried to use ErrorBoundary too, same error raised

**Reply by drarmstr** [reply]
Please clarify where you attempted to use an `<ErrorBoundary>`.

**Reply by babyfish-ct** [reply]
Sorry, here is the newest code

```
import {RecoilRoot, selector, useRecoilValueLoadable} from 'recoil';
import {ErrorBoundary} from 'react-error-boundary';
import { Suspense } from 'react';

const asyncError = selector({
  key: 'asyncError',
  get: async() => { throw new Error("Test Error"); }
});

function Test() {
  const loadable = useRecoilValueLoadable(asyncError);
  return (
    <h1>
      { loadable?.state === 'hasError' ? "Has error" : "No error" }
    </h1>
  );
}

function App() {
  return (
    <RecoilRoot>
      <ErrorBoundary fallback={<div>Error</div>}>
        <Suspense fallback={<div>Busy...</div>}>
          <Test/>
        </Suspense>
      </ErrorBoundary>
    </RecoilRoot>
  );
}

export default App;
```

The result is
![image](https://user-images.githubusercontent.com/15168900/116548112-554ea180-a926-11eb-8576-267213fbdbe0.png)


**I believe I've made some mistake... but I don't know where it is**

**Reply by drarmstr** [reply]
When I tried that code it seems to work as expected: https://codesandbox.io/s/ecstatic-boyd-xmzqb

In this case you are catching the error with the loadable, so it just shows "Has error" and the `<ErrorBoundary>` is not used.  If you used the `useRecoilValue()` hook instead of `useRecoilValueLoadable()` then it should throw and use the `<ErrorBoundary>` to catch it: https://codesandbox.io/s/gifted-tereshkova-p40e5?file=/src/App.js

**Reply by babyfish-ct** [reply]
Yes, I opened your link “https://codesandbox.io/s/ecstatic-boyd-xmzqb”， it works

In that sandbox, the version or recoil is "0.1.2", I change it to be "0.2.0", same error raised.

is it a bug of "0.2.0"?

**Reply by babyfish-ct** [reply]
|Sandbox|Recoil-0.1.2|Recoil-0.1.3|Recoil-0.20|
|------|--------|----------|--------------|
|https://codesandbox.io/s/ecstatic-boyd-xmzqb|Successed|Sucessed|Failed|
|https://codesandbox.io/s/gifted-tereshkova-p40e5?file=/src/App.js|Failed|Failed|Failed|

**Reply by drarmstr** [reply]
Hmm, yes, that looks like an issue then.  Let's open an issue on that; thanks!  cc @csantos42

**Reply by csantos42** [reply]
Looking into it now! Thanks for reporting

**Reply by csantos42** [reply]
Fixed in #1017
